bluetooth: conn: move `bt_conn_lookup_index` to public header

JordanYates opened the pull request:
Move the `bt_conn_lookup_index` declaration to a public header. This function is essentially the inverse of `bt_conn_index`, which is already public.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -718,6 +718,20 @@ void bt_conn_foreach(enum bt_conn_type type,
  */
 struct bt_conn *bt_conn_lookup_addr_le(uint8_t id, const bt_addr_le_t *peer);
 
+/**
+ * @brief Look up an existing connection by index.
+ *
+ *  Look up an existing connection based on the connection index.
+ *
+ *  The caller gets a new reference to the connection object which must be
+ *  released with bt_conn_unref() once done using the object.
+ *
+ * @param index Connection index from @ref bt_conn_index
+ *
+ *  @return Connection object or NULL if not found.
+ */
```
```suggestion
/** @brief Look up an existing connection by index.
 *
 *  Look up an existing connection based on the connection index.
 *
 *  The caller gets a new reference to the connection object which must be
 *  released with bt_conn_unref() once done using the object.
 *
 *  @param index Connection index from @ref bt_conn_index
 *
 *  @return Connection object or NULL if not found.
 */
```

Small formatting changes to follow style in file

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -718,6 +718,20 @@ void bt_conn_foreach(enum bt_conn_type type,
+ *
+ *  @return Connection object or NULL if not found.
+ */
+struct bt_conn *bt_conn_lookup_index(uint8_t index);
```
From https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#defining-smaller-prs

> If introducing a new API, the PR must include an example usage of the API. This provides context to the reviewer and prevents submitting PRs with unused APIs.

Is there any sample or test that will use this? Given that it's an inverse of an existing API, it may not be needed

We should also consider adding this to the release notes as "new API function"

JordanYates replied:
Added to the release notes, the tests and samples would be the existing usage of the function in the host I suppose.
The function already exists, this is just adding docs and making it public.

rugeGerritsen commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -718,6 +718,20 @@ void bt_conn_foreach(enum bt_conn_type type,
 struct bt_conn *bt_conn_lookup_addr_le(uint8_t id, const bt_addr_le_t *peer);
 
+/**
+ * @brief Look up an existing connection by index.
```
Does this API have some concerns with regards to thread safety?

Let's say that the user at some point stores `index 1`. And then later calls `bt_conn_lookup_index(1)`. I assume it cannot know that the returned connection is the same connection as used last time.

Nevertheless, I'm sure there are good use cases for this API. But we should make this clear to the user.

JordanYates replied:
> Does this API have some concerns with regards to thread safety?

I can't see why the concerns would be any different to `bt_conn_index`.

rugeGerritsen replied:
I was just wondering: An index cannot be used as input to any of the Bluetooth APIs, so the user cannot do anything "wrong/unexpected". When the user gets a connection reference, this is no longer true

Thalley replied:
I think it's a fair point from @rugeGerritsen 

Our APIs use a conn pointer instead of an index for a reason, and there may be unknown side effects from the higher layers relying on the index instead of the point. 

That being said, I don't think this is actually any different than our APIs for 
```c
int bt_hci_get_conn_handle(const struct bt_conn *conn, uint16_t *conn_handle);
struct bt_conn *bt_hci_conn_lookup_handle(uint16_t handle);
```

So any issues/challenges that may or not be with this suggestion function, already exists.

Thalley replied:
> I would like to expand on my comment above. For the second use case, `bt_conn_lookup_index` is slightly awkward since it does not move the from the index into the conn pointer.
> 
> ```c
> /* Connection index with reference. */
> int conn_index;
> void connected(struct bt_conn conn) {
>     conn_index = bt_conn_index(bt_conn_ref(conn));
> }
> 
> void clear_conn_index() {
>     /* How do I get rid of the reference in `conn_index`? */
> 
>     /* The reference in conn_index is not moved here. */
>     struct bt_conn conn = bt_conn_lookup_index(conn_index);
> 
>     /* Awkwardly, we have to bend the rules and move the reference to conn so it has two. */
>     // static_analysis_move_reference(conn, conn_index);
>     /* Now there are two references inhabiting the variable conn. */
> 
>     bt_conn_unref(conn); /* Get rid of index reference. */
>     bt_conn_unref(conn); /* Get rid of conn reference. */
> }
> ```
> 
> To avoid this awkwardness in the API, we need to define at least one function that takes a conn_index and moves a reference. One "elegant" solution is to have an additional function `bt_conn_from_index(conn_index)`, which is like bt_conn_lookup_index, but takes and moves a reference. But I can see the confusion this creates.
> 
> Another solution is to make a `bt_conn_unref_index`.

Hmm, I'm not sure I agree with this. You do `conn_index = bt_conn_index(bt_conn_ref(conn));`, but why would you take a reference if you don't store the pointer (i.e. the reference)? If you just did `conn_index = bt_conn_index(conn);` then it would not be a problem. 

Indexes and conn references are quite separate IMO, but it does make it a bit asymmetrical if only one of the `index` function deals with references.

jhedberg replied:
There's a very specific use case why these functions exist. If we clearly document that, then we shouldn't need to play these philosophical games with reference counting. The original use case is to allow doing a 1:1 mapping of higher level per-connection protocol context, without any memory overhead (user data pointers or linked lists). The idea being that each higher level protocol implementation does something like this:
```c
static struct my_protocol_data {
	...
} my_contexts[CONFIG_BT_MAX_CONN];

static void connected(struct bt_conn *conn)
{
	struct my_protocol_data *data = my_contexts[bt_conn_index(conn)];

	...
}

static void disconnected(struct bt_conn *conn)
{
	struct my_protocol_data *data = my_contexts[bt_conn_index(conn)];

	..clean up data..
	memset(data, 0, sizeof(*data));
}
```

If we restrict the uses of these APIs to something like the above, there really shouldn't be any risk of the kinds of bugs that you worry about.

alwa-nordic replied:
@jhedberg That's reasonable. But `bt_conn_lookup_index` plays no role in this use case. Is this an argument to close this PR?

jhedberg replied:
> @jhedberg That's reasonable. But `bt_conn_lookup_index` plays no role in this use case. Is this an argument to close this PR?

As long as the "protocol data" contains information that the connection "is alive" the reverse lookup can still make sense and be useful. There are two users of it in the host right now, one in `gatt.c` and another in `hci_core.c`. The GATT one looks buggy to me. `bt_conn_lookup_index()` can return both NULL (if there are no previous references) or a disconnected connection so there must be some checks for this if you don't track it somehow separately. The one in `hci_core.c` should be correct, assuming a well behaving controller.

alwa-nordic replied:
> > @jhedberg That's reasonable. But `bt_conn_lookup_index` plays no role in this use case. Is this an argument to close this PR?
> 
> As long as the "protocol data" contains information that the connection "is alive" the reverse lookup can still make sense and be useful. There are two users of it in the host right now, one in `gatt.c` and another in `hci_core.c`. The GATT one looks buggy to me. `bt_conn_lookup_index()` can return both NULL (if there are no previous references) or a disconnected connection so there must be some checks for this if you don't track it somehow separately. The one in `hci_core.c` should be correct, assuming a well behaving controller.

I think we agree then. So long as the caller of `bt_conn_lookup_index` can prove the object it got that index from is still alive, then that's proof `bt_conn_lookup_index` returns that object. Tracking references is one type of proof, but I agree it's not the only one.

alwa-nordic replied:
To steer this thread back on track: Does anyone have a good suggestion for API documentation for the average user to make it easy to use the new API correctly?

Thalley replied:
Should/could we align this with `bt_conn_foreach`? 

`bt_conn_foreach` uses `bt_conn_ref` to determine if the `conn` object has a reference, i.e. it only iterates on those that have a non-0 reference counter. Should `bt_conn_lookup_index` return `NULL` if the reference counter is `0`? 

Do we want the possible return values of `bt_conn_lookup_index` and `bt_conn_foreach` to differ? 
If not, then I guess we could/should align the documentation of those 2 functions, as they effectively provide the same feature. 

That also reminded me that `bt_conn_lookup_index` should take the type into account as well. 
A ISO `bt_conn` object and an ACL `bt_conn` object may have the same index, so with the current suggested API we may have

```c
uint8_t index = bt_conn_index(iso_conn); /* returns 0 */
struct bt_conn *conn = bt_conn_lookup_index(index); /* returns ACL at index 0 */
```

This hasn't been a problem when used internally, as that has always been used in an ACL context, but that cannot be guaranteed in a API context

Thalley commented on doc/releases/release-notes-4.1.rst:
```diff
@@ -135,6 +135,10 @@ New APIs and options
     * :c:member:`bt_mesh_health_cli::update` callback can be used to periodically update the message
       published by the Health Client.
 
+  * Host
+
+    * :c:func:`bt_conn_lookup_index` is now available as a counterpart to :c:func:`bt_conn_index`.
```
The release note structure has changed for 4.1, and shouldn't really contain much description: 

> If you are describing a feature or functionality, consider adding it to the actual project documentation rather than the release notes, so that the information does not get lost in time.

To follow the non-BT mesh lists in this file, consider just adding it as 
```suggestion
  * Host

    * :c:func:`bt_conn_lookup_index`
```

I think generally we put `host` before `mesh` in the release notes like I've done here: https://github.com/zephyrproject-rtos/zephyr/pull/85287/files#diff-751bf4238d5b9251c91a0cde20653713228079e389f60dd6221bd3b596ed8c61

May be a good idea to do that here as well, so that we get a merge conflict that can be properly addressed rather than ending up with 2 host entries

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -718,6 +718,19 @@ void bt_conn_foreach(enum bt_conn_type type,
+ *  The caller gets a new reference to the connection object which must be
+ *  released with bt_conn_unref() once done using the object.
+ *
+ *  @param index Connection index from @ref bt_conn_index
```
```suggestion
 *  @param index Connection index from bt_conn_index()
```

JordanYates committed:
bluetooth: conn: move `bt_conn_lookup_index`` to public header

Move the `bt_conn_lookup_index` declaration to a public header. This
function is essentially the inverse of `bt_conn_index`, which is already
public.

Signed-off-by: Jordan Yates <jordan@embeint.com>

Thalley reviewed (changes requested):
I'd like to see the suggest missing documentation from https://github.com/zephyrproject-rtos/zephyr/pull/85546#discussion_r1950686216 being added. 

A conn object returned by this function my not be connected nor contain valid values, which should be made clear to the user

JordanYates commented:
Who knew this internal function was so problematic :)
I'm still not clear on what is actually desired from a documentation perspective.

jhedberg commented:
> A conn object returned by this function my not be connected nor contain valid values

It may indeed not be connected, but it *should* contain valid values since the API only returns objects with a pre-existing refcount > 0

Thalley commented:
> Who knew this internal function was so problematic :) I'm still not clear on what is actually desired from a documentation perspective.

My perspective: 
1) It need to also include the type of connection we are looking up. We have several arrays of connections and the index is only unique per type. 
2) The documentation of `bt_conn_lookup_index` should match the documentation of `bt_conn_foreach`; they are effectively doing the same, except that `bt_conn_foreach` will run on all possible IDs. 
3) To ensure that `bt_conn_foreach` and `bt_conn_lookup_index` can share similar documentation, their behavior should be similar. I suggest to achieve this by modifying `bt_conn_foreach` to call  `bt_conn_lookup_index` for each index. 

If any additional documentation is required, that should also be added to `bt_conn_foreach`

JordanYates commented:
> > Who knew this internal function was so problematic :) I'm still not clear on what is actually desired from a documentation perspective.
> 
> My perspective:
> 
>     1. It need to also include the type of connection we are looking up. We have several arrays of connections and the index is only unique per type.
> 
>     2. The documentation of `bt_conn_lookup_index` should match the documentation of `bt_conn_foreach`; they are effectively doing the same, except that `bt_conn_foreach` will run on all possible IDs.
> 
>     3. To ensure that `bt_conn_foreach` and `bt_conn_lookup_index` can share similar documentation, their behavior should be similar. I suggest to achieve this by modifying `bt_conn_foreach` to call  `bt_conn_lookup_index` for each index.
> 
> 
> If any additional documentation is required, that should also be added to `bt_conn_foreach`

That's all fair enough but is a pretty big explosion of scope from just moving a function from an internal to a public header.
Quite frankly this is a low value PR for me and I'm not going to get any value from the time spent pushing the changes.

Thalley commented:
> > > Who knew this internal function was so problematic :) I'm still not clear on what is actually desired from a documentation perspective.
> > 
> > 
> > My perspective:
> > ```
> > 1. It need to also include the type of connection we are looking up. We have several arrays of connections and the index is only unique per type.
> > 
> > 2. The documentation of `bt_conn_lookup_index` should match the documentation of `bt_conn_foreach`; they are effectively doing the same, except that `bt_conn_foreach` will run on all possible IDs.
> > 
> > 3. To ensure that `bt_conn_foreach` and `bt_conn_lookup_index` can share similar documentation, their behavior should be similar. I suggest to achieve this by modifying `bt_conn_foreach` to call  `bt_conn_lookup_index` for each index.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If any additional documentation is required, that should also be added to `bt_conn_foreach`
> 
> That's all fair enough but is a pretty big explosion of scope from just moving a function from an internal to a public header. Quite frankly this is a low value PR for me and I'm not going to get any value from the time spent pushing the changes.

That's fair :) Feel free to create a GH issue requesting this to become public, and then we can be aware of the use case and request and then use that to (hopefully) implement this in the future

JordanYates commented:
> That's fair :) Feel free to create a GH issue requesting this to become public, and then we can be aware of the use case and request and then use that to (hopefully) implement this in the future.

My interpretation of the blockers here was that the existing documentation (and implementation?) of the (already public) `bt_conn_foreach` is insufficient. If that was not the case, then copying the docs from `bt_conn_foreach` as requested would be simple.

Thalley commented:
> My interpretation of the blockers here was that the existing documentation (and implementation?) of the (already public) `bt_conn_foreach` is insufficient.

You can indeed use `bt_conn_foreach` and `bt_conn_index` to do the same, although in a less optimal way as you'd always iterate on all connections
